modules/aggregation-layers/src/screen-grid-layer/screen-grid-layer.ts

```diff
@@ -138,208 +111,182 @@ export type _ScreenGridLayerProps<DataT> = {
+    super.updateState(params);
+
+    const {props, oldProps, changeFlags} = params;
+    const {cellSizePixels, aggregation} = props;
+    if (
+      changeFlags.dataChanged ||
+      changeFlags.updateTriggersChanged ||
+      changeFlags.viewportChanged ||
+      aggregation !== oldProps.aggregation ||
+      cellSizePixels !== oldProps.cellSizePixels
+    ) {
+      const {width, height} = this.context.viewport;
+      const {aggregator} = this.state;
+
+      if (aggregator instanceof GPUAggregator) {
+        aggregator.setProps({
+          binIdRange: [
+            [0, Math.ceil(width / cellSizePixels)],
+            [0, Math.ceil(height / cellSizePixels)]
+          ]
+        });
+      }
+
+      aggregator.setProps({
+        pointCount: this.getNumInstances(),
+        operations: [aggregation],
+        binOptions: {
+          cellSizePixels
+        }
+      });
+    }
+
+    if (changeFlags.viewportChanged) {
+      // Rerun aggregation on viewport change
+      this.state.aggregator.setNeedsUpdate();
+    }
   }
 
-  renderLayers(): LayersList | Layer {
-    if (!this.state.supported) {
-      return [];
+  onAttributeChange(id: string) {
+    const {aggregator} = this.state;
+    switch (id) {
+      case 'positions':
+        aggregator.setNeedsUpdate();
+        break;
+
+      case 'counts':
+        aggregator.setNeedsUpdate(0);
+        break;
+
+      default:
+      // This should not happen
     }
-    const {maxTexture, numRow, numCol, weights} = this.state;
-    const {updateTriggers} = this.props;
-    const {aggregationBuffer} = weights.count;
+  }
+
+  renderLayers(): LayersList | Layer | null {
+    const {aggregator} = this.state;
     const CellLayerClass = this.getSubLayerClass('cells', ScreenGridCellLayer);
+    const binAttribute = aggregator.getBins();
+    const weightAttribute = aggregator.getResult(0);
 
     return new CellLayerClass(
       this.props,
       this.getSubLayerProps({
-        id: 'cell-layer',
-        updateTriggers
+        id: 'cell-layer'
       }),
       {
-        data: {attributes: {instanceCounts: aggregationBuffer}},
-        maxTexture,
-        numInstances: numRow * numCol
+        data: {
+          length: aggregator.numBins,
+          attributes: {
+            getBin: binAttribute,
+            getWeight: weightAttribute
+          }
+        },
+        // Data has changed shallowly, but we likely don't need to update the attributes
+        dataComparator: (data, oldData) => data.length === oldData.length,
+        updateTriggers: {
+          getBin: binAttribute?.buffer,
+          getWeight: weightAttribute?.buffer
+        },
+        parameters: {
+          depthWriteEnabled: false,
+          ...this.props.parameters
+        },
+        // Evaluate domain at draw() time
+        colorDomain: () => this.props.colorDomain || aggregator.getResultDomain(0),
+        // Extensions are already handled by the GPUAggregator, do not pass it down
```

Comment:
For ScreenGridLayer there's no existing extension that should be applied to the cells directly, and unlikely for any custom extensions.

This will be tricky for GridLayer/HexagonLayer. FillStyleExtension and TerrainExtension should affect the screen pass only, and *FilterExtension should affect the aggregation pass only.

We could add some flag to the extensions to categorize them?

Reply:
Would it be sufficient for the extension to signal which stage of the GPU pipeline it affects? E.g. `'vertex'`/`'fragment'`. The aggregation pass should only use the `'vertex'` extensions and the screen pass the `'fragment'`.

Even so, how does this work for `DataFilterExtension`? There we would like the vertex shader injections to hide the filtered data points before aggregation, but in the screen pass the `dataFilter_value` needs to both shrink the geometry in the vertex stage & fade in the fragment stage

An alternative would be to have the extensions understand the concept of what context they are being invoked in, by looking at what type of `Pass` they are being rendered in. The `DataFilterExtension` could then do one thing in the `AggregationPass` and another in the `ScreenPass`